# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionImpl.h
# Commit: fb1a765ec213
# Full Hash: fb1a765ec213b12a3c6aa1474e8dc42585b1c2aa
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2022-12-21 21:21:23
# Regressor Bug: 1800136
# File Overlap Count: 1
# Description:
#   Bug 1800136: Fix gap between PC.close() and the storage of its final stats in PeerConnectionCtx. r=ng
#   
#   Ensures that:
#   1) We never have stats report in a still-registered PC _and_ the stats for
#      closed PCs.
# ==============================================================================

diff -r 3621f5bf1cb0 -r fb1a765ec213 dom/media/webrtc/jsapi/PeerConnectionImpl.h
--- a/dom/media/webrtc/jsapi/PeerConnectionImpl.h	Tue Dec 20 22:12:10 2022 +0000
+++ b/dom/media/webrtc/jsapi/PeerConnectionImpl.h	Tue Dec 20 22:12:10 2022 +0000
@@ -467,6 +467,7 @@
   void OnSetDescriptionError();
 
   bool IsClosed() const;
+
   // called when DTLS connects; we only need this once
   nsresult OnAlpnNegotiated(bool aPrivacyRequested);
 
@@ -523,6 +524,14 @@
 
   void SendWarningToConsole(const nsCString& aWarning);
 
+  const UniquePtr<dom::RTCStatsReportInternal>& GetFinalStats() const {
+    return mFinalStats;
+  }
+
+  void DisableLongTermStats() { mDisableLongTermStats = true; }
+
+  bool LongTermStatsIsDisabled() const { return mDisableLongTermStats; }
+
  private:
   virtual ~PeerConnectionImpl();
   PeerConnectionImpl(const PeerConnectionImpl& rhs);
@@ -539,6 +548,7 @@
                                      uint32_t aMaxMessageSize, bool aMMSSet);
 
   nsresult CheckApiState(bool assert_ice_ready) const;
+  void StoreFinalStats(UniquePtr<dom::RTCStatsReportInternal>&& report);
   void CheckThread() const { MOZ_ASSERT(NS_IsMainThread(), "Wrong thread"); }
 
   // test-only: called from AddRIDExtension and AddRIDFilter
@@ -656,6 +666,13 @@
   bool mCallTelemStarted = false;
   bool mCallTelemEnded = false;
 
+  // We _could_ make mFinalStatsQuery be an RTCStatsReportPromise, but that
+  // would require RTCStatsReportPromise to no longer be exclusive, which is
+  // a bit of a hassle, and not very performant.
+  RefPtr<GenericNonExclusivePromise> mFinalStatsQuery;
+  UniquePtr<dom::RTCStatsReportInternal> mFinalStats;
+  bool mDisableLongTermStats = false;
+
   // Start time of ICE.
   mozilla::TimeStamp mIceStartTime;
   // Hold PeerConnectionAutoTimer instances for each window.